R/write.report.R

```diff
@@ -21,156 +20,159 @@
+      for (.model in names(x[[.scenario]])) {
+        scenario <- ifelse(is.null(scenarioCall), .scenario, scenario)
+        model <- ifelse(is.null(modelCall), .model, model)
+        write.report(x[[.scenario]][[.model]], file = file, model = model, scenario = scenario, unit = unit,
+                     ndigit = ndigit, append = append, skipempty = skipempty, extracols = extracols)
+        append <- TRUE
+      }
     }
   } else {
-
-    if(!is.magpie(x)) stop("Input is not a MAgPIE object!")
-    dimnames(x)[[1]] <- sub("^GLO(\\.{0,1}[0-9]*)$","World\\1",dimnames(x)[[1]])
-    
-    # If data was read in by read.report there is an attribute $dimnames$scenario.model.variable.
-    # This will be used below to structure the output of write.report exactly like the input.
-    # That means: write.report automatically recognizes models and scenarios and 
-    # does not put it into the data-dimension
-    if (scenario[1] == "default" & length(scenario) == 1 & length(names(attr(x,"dimnames"))) >2) {
-      if (names(attr(x,"dimnames"))[[3]] == "scenario.model.variable") {
-        scenario <- fulldim(x)[[2]]$scenario
-        model <- fulldim(x)[[2]]$model
+    if (!is.magpie(x)) stop("Input is not a MAgPIE object!")
+    x <- prepareData(x, model = model, scenario = scenario, unit = unit, skipempty = skipempty,
+                     ndigit = ndigit, extracols = extracols)
+    if (is.null(file)) return(x)
+    if (!file.exists(file)) append <- FALSE
+    if (append) {
+      # check header for consistency
+      header <- read.table(file, nrows = 1, sep = ";", stringsAsFactors = FALSE)
+      years1 <- as.numeric(header[sapply(header, is.numeric)]) # nolint
+      years2 <- as.numeric(colnames(x)[!is.na(suppressWarnings(as.numeric(colnames(x))))])
+      union <- sort(union(years1, years2))
+      addycols <- function(data, years) {
+        ycols <- !is.na(suppressWarnings(as.numeric(colnames(data))))
+        tmp <- data[ycols]
+        data <- data[!ycols]
+        data[as.character(sort(years))] <- "N/A"
+        data[names(tmp)] <- tmp
+        return(data)
+      }
+      if (length(union) > length(years1)) {
+        data <- read.table(file, sep = ";", stringsAsFactors = FALSE, header = TRUE, check.names = FALSE)
+        data <- data[-length(data)]
+        write.table(addycols(data, union), file, quote = FALSE, sep = ";", row.names = FALSE,
+                    col.names = TRUE, append = FALSE, eol = ";\n")
+      }
+      if (length(union) > length(years2)) {
+        x <- addycols(as.data.frame(x), union)
       }
     }
+    write.table(x, file, quote = FALSE, sep = ";", row.names = FALSE, col.names = !append,
+                append = append, eol = ";\n")
+  }
+}
+
+prepareData <- function(x, model = NULL, scenario = NULL, unit = NULL, skipempty = FALSE,
+                        ndigit = 4, extracols = NULL) {
+  if (!requireNamespace("reshape2", quietly = TRUE)) stop("The package reshape2 is required for write.report!")
```

Comment:
It's a little confusing that this check is not done in write.report directly, although the message suggests that. Maybe move the check to write.report?

Reply:
In this case I disagree as moving it to `write.report` would make it harder to figure out why `reshape2` is required as the function call is hidden in `prepareData` and it would increase the risk that this check would still remain in the code even if the reshape2-dependency would be removed in `prepareData`